api/agent/agent.go

```diff
@@ -278,55 +284,88 @@ func (a *agent) getSlot(ctx context.Context, call *call) (Slot, error) {
+// containers if needed. Upon shutdown or activity timeout, hotLauncher exits and during exit,
+// it destroys the slot queue.
+func (a *agent) hotLauncher(ctx context.Context, call *call) {
+
+	// Let use 60 minutes or 2 * IdleTimeout as hot queue idle timeout, pick
+	// whichever is longer. If in this time, there's no activity, then
+	// we destroy the hot queue.
+	timeout := time.Duration(60) * time.Minute
+	idleTimeout := time.Duration(call.IdleTimeout) * time.Second * 2
+	if timeout < idleTimeout {
+		timeout = idleTimeout
+	}
 
+	logger := common.Logger(ctx)
+	logger.WithField("launcher_timeout", timeout).Info("Hot function launcher starting")
 	isAsync := call.Type == models.TypeAsync
 
-launchLoop:
+	defer func() {
+		logger.Info("Hot function launcher terminating")
+		a.slotMgr.destroySlotQueue(call.slots)
+		a.wg.Done()
+	}()
+
 	for {
-		// Check/evaluate if we need to launch a new hot container
-		doLaunch, stats := call.slots.isNewContainerNeeded()
-		common.Logger(ctx).WithField("stats", stats).Debug("checking hot container launch ", doLaunch)
+		select {
+		case <-a.shutdown: // server shutdown
+			return
+		case <-time.After(timeout):
+			if call.slots.isIdle() {
+				logger.Info("Hot function launcher timed out")
+				return
+			}
+		case <-call.slots.signaller:
+		}
 
-		if doLaunch {
-			ctxToken, tokenCancel := context.WithCancel(context.Background())
+		isNeeded, stats := call.slots.isNewContainerNeeded()
+		logger.WithField("stats", stats).Debug("Hot function launcher stats")
+		if !isNeeded {
+			continue
+		}
 
-			// wait on token/slot/timeout whichever comes first
-			select {
-			case tok, isOpen := <-a.resources.GetResourceToken(ctxToken, call.Memory, isAsync):
-				tokenCancel()
-				if !isOpen {
-					return nil, models.ErrCallTimeoutServerBusy
-				}
-				go a.runHot(ctx, call, tok)
-			case s, ok := <-call.slots.getDequeueChan():
-				tokenCancel()
-				if !ok {
-					return nil, errors.New("slot shut down while waiting for hot slot")
-				}
-				if s.acquireSlot() {
-					if s.slot.Error() != nil {
-						s.slot.Close()
-						return nil, s.slot.Error()
-					}
-					return s.slot, nil
-				}
+		resourceCtx, cancel := context.WithCancel(context.Background())
+		logger.WithField("stats", stats).Info("Hot function launcher starting hot container")
 
-				// we failed to take ownership of the token (eg. container idle timeout)
-				// try launching again
-				continue launchLoop
-			case <-ctx.Done():
-				tokenCancel()
-				return nil, ctx.Err()
+		select {
+		case tok, isOpen := <-a.resources.GetResourceToken(resourceCtx, call.Memory, isAsync):
+			cancel()
+			if isOpen {
+				go a.runHot(ctx, call, tok)
+			} else {
+				// this means the resource was impossible to reserve (eg. memory size we can never satisfy)
+				call.slots.queueSlot(&hotSlot{done: make(chan struct{}), err: models.ErrCallTimeoutServerBusy})
 			}
+		case <-time.After(timeout):
+			cancel()
+			if call.slots.isIdle() {
```

Comment:
is there a race here where a call could come in where this is shutting down and this thread wouldn't get re-launched?

Reply:
nice catch, I can fix this up by increasing the scope of the 'waiting' state in:

https://github.com/skinowski/fn/blob/f92901e673af9c3b0d9c271fa46fee7169afc22e/api/agent/agent.go#L277

I'll move this before the lookup to cover it. If it times out when someone is about lookup the hot queue map, the slot queue won't be idle.